bccsp/pkcs11/pkcs11.go

```diff
@@ -84,25 +84,44 @@ func loadLib(lib, pin, label string) (*pkcs11.Ctx, uint, *pkcs11.SessionHandle,
 func (csp *impl) getSession() (session pkcs11.SessionHandle) {
 	select {
 	case session = <-csp.sessions:
-		logger.Debugf("Reusing existing pkcs11 session %+v on slot %d\n", session, csp.slot)
+		_, err := csp.ctx.GetSessionInfo(session)
+		if err != nil {
+			logger.Warningf("Get session info failed [%s], getting new session\n", err)
+			session = csp.createSession()
+		} else {
+			logger.Debugf("Reusing existing pkcs11 session %+v on slot %d\n", session, csp.slot)
+		}
 
 	default:
 		// cache is empty (or completely in use), create a new session
-		var s pkcs11.SessionHandle
-		var err error
-		for i := 0; i < 10; i++ {
-			s, err = csp.ctx.OpenSession(csp.slot, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
-			if err != nil {
-				logger.Warningf("OpenSession failed, retrying [%s]\n", err)
-			} else {
-				break
-			}
-		}
+		session = csp.createSession()
+	}
+	return session
+}
+
+func (csp *impl) createSession() pkcs11.SessionHandle {
+	var s pkcs11.SessionHandle
+	var err error
+	for i := 0; i < 10; i++ {
+		s, err = csp.ctx.OpenSession(csp.slot, pkcs11.CKF_SERIAL_SESSION|pkcs11.CKF_RW_SESSION)
 		if err != nil {
-			panic(fmt.Errorf("OpenSession failed [%s]", err))
+			logger.Warningf("OpenSession failed, retrying [%s]\n", err)
+		} else {
+			break
+		}
+	}
```

Comment:
Iterating in a tight loop is rarely a good idea. (It was bad where you copied it from and it's bad here.)

If it's not successful on the first attempt, it's unlikely it's going to be immediately accessible on the second. This should probably be an exponential backoff.

Reply:
@sykesm thanks for the feedback. I did not intent to change the functionality of creation of session. So I retained the code as is, just moved it to a new function.

The intent of this PR is to check if a session is invalid and if we should be getting a new one.